core/unit_test/TestGraph.hpp

```diff
@@ -739,4 +739,61 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), end_of_submit_control_flow) {
             value_A + 2 * value_B + value_C + value_D + value_E + value_F);
 }
 
+template <typename Exec>
+struct GraphNodeTypes {
+  //! Type of a kernel node built using a Kokkos parallel construct.
+  using kernel_t =
+      Kokkos::Impl::GraphNodeKernelImpl<Exec, Kokkos::RangePolicy<Exec>,
+                                        CountTestFunctor<Exec>,
+                                        Kokkos::ParallelForTag>;
+
+  //! Type of an aggregate node.
+  using aggregate_t = typename Kokkos::Impl::GraphImpl<Exec>::aggregate_impl_t;
+};
+
+TEST(TEST_CATEGORY, is_graph_kernel_v) {
+  using types = GraphNodeTypes<TEST_EXECSPACE>;
+
+  static_assert(Kokkos::Impl::is_graph_kernel_v<types::kernel_t>);
+  static_assert(!Kokkos::Impl::is_graph_kernel_v<types::aggregate_t>);
+}
+
+// This test checks the node types before/after a 'when_all'.
+TEST(TEST_CATEGORY, when_all_type) {
+  using kernel_functor_t = CountTestFunctor<TEST_EXECSPACE>;
+  using graph_t          = Kokkos::Experimental::Graph<TEST_EXECSPACE>;
+  using graph_impl_t     = Kokkos::Impl::GraphImpl<TEST_EXECSPACE>;
+  using root_t =
+      Kokkos::Experimental::GraphNodeRef<TEST_EXECSPACE,
+                                         Kokkos::Experimental::TypeErasedTag,
+                                         Kokkos::Experimental::TypeErasedTag>;
+  using node_kernel_impl_t = Kokkos::Impl::GraphNodeKernelImpl<
+      TEST_EXECSPACE,
+      Kokkos::RangePolicy<TEST_EXECSPACE,
+                          Kokkos::Impl::IsGraphKernelTag>,
+      kernel_functor_t, Kokkos::ParallelForTag>;
+  using first_layer_t =
```

Comment:
Such variables could also be named `node_ref_root_t`, `node_ref_first_layer_t`, `node_ref_agg_t` and `node_ref_tail_t`. It's fine to keep it as is too.

Reply:
Done.